pkg/kubelet/pod_container_deletor.go

```diff
@@ -80,20 +81,27 @@ func getContainersToDeleteInPod(filterContainerId string, podStatus *kubecontain
 
 	// Find the exited containers whose name matches the name of the container with id being filterContainerId
 	var candidates containerStatusbyCreatedList
+	var infraCandidates containerStatusbyCreatedList
 	for _, containerStatus := range podStatus.ContainerStatuses {
 		if containerStatus.State != kubecontainer.ContainerStateExited {
 			continue
 		}
 		if matchedContainer == nil || matchedContainer.Name == containerStatus.Name {
-			candidates = append(candidates, containerStatus)
+			if containerStatus.Name == leaky.PodInfraContainerName {
+				infraCandidates = append(infraCandidates, containerStatus)
+			} else {
+				candidates = append(candidates, containerStatus)
+			}
 		}
 	}
 
 	if len(candidates) <= containersToKeep {
-		return containerStatusbyCreatedList{}
+		// Always destroy infra containers to ensure networking gets torn down
```

Comment:
There are actually two different moments which may trigger network teardown:

(1) when infra container is down
(2) when infra container is removed

I think codes in `removeContainer()` handles (2) and here is a fix for (1).

To reduce IPAM leaking asap, I suggest we only trigger network tear down at infra container stopped, e.g. perform infra container network teardown after getting evictableContainers (and before removing any containers).